subcommand/mesh-init/command.go

```diff
@@ -171,6 +181,69 @@ func (c *Command) loginToAuthMethod(tokenFile string, taskMeta awsutil.ECSTaskMe
 	}, backoff.NewConstantBackOff(2*time.Second), retryLogger(c.log))
 }
 
+func (c *Command) waitForTokenReplication(cfg *api.Config) error {
+	// A workaround to check that the ACL token is replicated to other Consul servers.
+	// Code borrowed from: https://github.com/hashicorp/consul-k8s/pull/887
+	//
+	// This problem can potentially occur because of:
+	//
+	// - Replication lag: After a token is created on the Consul server leader it may take up to
+	//   100ms (typically) for the token to be replicated to server followers:
+	//   https://www.consul.io/docs/install/performance#read-write-tuning
+	// - Stale consistency mode: Consul clients may connect to a Consul server follower, which may
+	//   have stale state, in order to reduce load on the server leader.
+	// - Negative caching: When a Consul server validates a token, if the server does know about the
+	//   token (e.g. due to replication lag), then an "ACL not found" response is cached. By default,
+	//   the cache time is 30s: https://www.consul.io/docs/agent/config/config-files#acl_token_ttl.
+	// - Sticky connections: Consul clients maintain a connection to a single Consul server, and
+	//   these connections are only rebalanced every 2-3 mins.
+	//
+	// Therefore, an "ACL not found" error may be cached just after token creation. When this
+	// happens, the token will be unusable for the acl_token_ttl (30s by default). Retrying requests
+	// won't help since client likely won't change Consul servers for a potentially longer time
+	// (2-3 min). If you are running 3 Consul servers, you have a 2/3 chance to hit a follower and
+	// encounter this problem, so this is a potentially frequent problem.
+	//
+	// We don't want to delay start up by the "long" cache time (default 30s). Instead, we wait
+	// for the token to be read successfully in stale consistency mode, which should take <=100ms since
+	// that is the typical Raft replication time.
+	//
+	// The does not eliminate this problem completely. It's still possible for this call and the
+	// next call to reach different servers and those servers to have different states from each
+	// other, but this is unlikely since clients use sticky connections.
+
+	// Mesh-init talks to the local Consul client agent (for now). We need this to hit the Consul
+	// server(s) directly.
+	newCfg := api.DefaultConfig()
+	newCfg.Address = c.config.ConsulHTTPAddr
+	newCfg.TLSConfig.CAFile = c.config.ConsulCACertFile
+	newCfg.TokenFile = cfg.TokenFile
+	newCfg.Token = cfg.Token
```

Comment:
I don't love the way I "clone" this config. I could change the function to receive a value/copy, e.g. `waitForTokenReplication(cfg api.Config)`, so that the function receives a copy, but there are some pointer fields on the `api.Config` type...

Maybe a common function to construct a new config will help.

Reply:
I updated this. I realized it just needed the TokenFile and not the Token as well.